clients/pkg/promtail/targets/heroku/target.go

```diff
@@ -0,0 +1,177 @@
+package heroku
+
+import (
+	"flag"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	herokuEncoding "github.com/heroku/x/logplex/encoding"
+	"github.com/imdario/mergo"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/model/relabel"
+	"github.com/weaveworks/common/server"
+
+	"github.com/grafana/loki/clients/pkg/promtail/api"
+	"github.com/grafana/loki/clients/pkg/promtail/scrapeconfig"
+	"github.com/grafana/loki/clients/pkg/promtail/targets/target"
+
+	"github.com/grafana/loki/pkg/logproto"
+	util_log "github.com/grafana/loki/pkg/util/log"
+)
+
+type Target struct {
+	logger         log.Logger
+	handler        api.EntryHandler
+	config         *scrapeconfig.HerokuTargetConfig
+	jobName        string
+	server         *server.Server
+	metrics        *Metrics
+	relabelConfigs []*relabel.Config
+}
+
+func NewTarget(metrics *Metrics, logger log.Logger, handler api.EntryHandler, jobName string, config *scrapeconfig.HerokuTargetConfig, relabel []*relabel.Config) (*Target, error) {
+
+	ht := &Target{
+		metrics:        metrics,
+		logger:         logger,
+		handler:        handler,
+		jobName:        jobName,
+		config:         config,
+		relabelConfigs: relabel,
+	}
+
+	// Bit of a chicken and egg problem trying to register the defaults and apply overrides from the loaded config.
+	// First create an empty config and set defaults.
+	defaults := server.Config{}
+	defaults.RegisterFlags(flag.NewFlagSet("empty", flag.ContinueOnError))
+	// Then apply any config values loaded as overrides to the defaults.
+	if err := mergo.Merge(&defaults, config.Server, mergo.WithOverride); err != nil {
+		level.Error(logger).Log("msg", "failed to parse configs and override defaults when configuring push server", "err", err)
```

Comment:
Agree, should return an error since it's a configuration failure. I copied the configuration logic from [the push target](https://github.com/grafana/loki/blob/main/clients/pkg/promtail/targets/lokipush/pushtarget.go#L57) which does almost (if not) the same. I could also extract the shared logic. Wdyt?

Reply:
Just dived a little deeper into that mergo implementation and couldn't cause an error. If you look at [the implementation](https://github.com/grafana/loki/blob/6fde88ef92378778c1c50126fda4fa2e03bcaf9d/vendor/github.com/imdario/mergo/merge.go#L345-L367) everything ends up playing nicely since both merged values are fixed, and according to the assumptions mergo makes for working correctly. My guess is that log was added to avoid not handling the error.

Reply:
I'm a little hesitant to rely on the internal (opaque) behaviour of a library which could change at some point.
Can we fail this function with that error in this case please? Either way it shouldn't change things as they are now, but it will handle the situation gracefully & predictably if `mergo` ever changes or is replaced by something else.

Reply:
Done!